docs: add info about request timeouts

bob-watson opened the pull request:
Docs-only change that:

* Adds a short note about how service workers report request timeouts
* Updated heading levels to present a clearer hierarchy and cleaner local TOC
* Removed documentation lint errors
* Fixes #46445

thePunderWoman committed:
refactor(platform-browser): Updates keyboard event library to support code field (#46030)

On MacOS, pressing the alt key and another key turns into symbols, and
doesn't match the intended behavior. For example, `keydown.alt.s`
reports instead as `keydown.alt.ß`. We rely on the `key` field and not
the `code` field, which properly reports the code for S in this case.

This change adds support to allow users to specify they want to look at
the `code` field instead of the `key` field within their event string.
Example: `keydown.code.alt.leftshift` would only match the LeftShift and
not the right shift based on code values. It would also allow the user
to specify `keydown.code.alt.keys` to match S instead of ß when alt /
option is pressed on MacOS and would also work on Windows.

Fixes #45992

PR Close #46030

cexbrayat committed:
feat(forms): add `FormBuilder.record()` method (#46485)

The new `FormRecord` entity introduced in Angular v14 does not have its builder method.
This commit adds it, allowing to write:

```
const fb = new FormBuilder();
fb.record({ a: 'one' });
```

This works for both the `FormBuilder` and the `NonNullableFormBuilder`

PR Close #46485

dario-piotrowicz committed:
fix(docs-infra): convert docs select for versions into navigation (#46674)

convert the select for the docs versions into a proper navigation to
make it more clear for users and also to improve its accessibility

resolves #44339

PR Close #46674

angular-robot committed:
build: update dependency karma-jasmine to v5

| datasource | package       | from  | to    |
| ---------- | ------------- | ----- | ----- |
| npm        | karma-jasmine | 4.0.2 | 5.1.0 |

devversion committed:
fix(core): do not invoke jasmine `done` callback multiple times with `waitForAsync`

Currently tests written using `waitForAsync` would be prone to Jasmine
warnings or errors (depending on the version) for tests incorrectly
invoking asynchronous jasmine `done` callbacks multiple times.

This can happen because the async test zone logic schedules the
`done` callback to be called using `setTimeout`, but this could
be invoked multiple times, causing multiple `done` invocations to
be scheduled. Most of the issues have been resolved with #45025,
but it does not solve the case of multiple tasks finished and callbacks
being scheduled.

Technically, the current logic is built in way that _should_ result in
`_finishCallbackIfDone` and eventually the `done` callback to be invoked
at maximium once. This is unfortunately not the case in some rather
advanced/unexpected scenarios (like our AngularJS upgrade tests) where
the scenario is the following (and microtasks from before the actual
`waitForAsync` spec are still completing -- which is valid):

```
1. A test `beforeEach` schedules a microtask in the ProxyZone.
2. An actual empty `it` spec executes in the AsyncTestZone` (using e.g. `waitForAsync`).
3. The `onInvoke` invokes `_finishCallbackIfDone` because the spec runs synchronously.
4. We wait the scheduled timeout (see below) to account for unhandled promises.
5. The microtask from (1) finishes and `onHasTask` is invoked.

--> We register a second `_finishCallbackIfDone` even though we have scheduled a timeout.
--> we execute the `done` callback twice because the async zone spec state is "stable"
```

devversion committed:
test: update tests to account for `karma-jasmine` v5.0.0

Karma jasmine updated the `jasmine-core` dependency. Jasmine is now more
strict when:

* The done callback is invoked multiple times
* The done callback is used, while a promise is also returned
* The done callback is treated as error when e.g. a number is returned
  as first argument. This was the case with `requestAnimationFrame`.

devversion committed:
test: fix `fake-async-test` spec to properly test prefixed animation frame functions

In the fake async test for `zone.js` there is some logic to test various
prefixes of the request animation frame functions. This logic does not
have any effect currently, except for generating the same test at
maximum three times.

This commit fixes the test to actually test what it meant to do
initially.

pkozlowski-opensource committed:
feat(core): introduce createApplication API (#46475)

The `createApplication` function makes it possible to create an
application instance (represented by the `ApplicationRef`)
without bootstrapping any components. It is useful in the
situations where ones wants to decouple and delay components
rendering and / or render multiple root components in one
application. Angular elements can use this API to create
custom element types with an environment linked to a
created application.

PR Close #46475

pkozlowski-opensource committed:
test(elements): add a test for standalone element creation (#46475)

This commit adds a test demonstrating how a standalone Angular component
could be turned into a custom element.

PR Close #46475

Commit:
docs: add info about request timeouts

* Adds a short note about how service workers report request timeouts
* Updated heading levels to present a clearer hierarchy and cleaner local TOC
* Removed documentation lint errors

fixes #46445

Commit:
docs: add info about request timeouts

* Adds a short note about how service workers report request timeouts
* Updated heading levels to present a clearer hierarchy and cleaner local TOC
* Removed documentation lint errors

fixes #46445

mary-poppins commented:
You can preview c04ab73 at https://pr46803-c04ab73.ngbuilds.io/.
You can preview 1804e4e at https://pr46803-1804e4e.ngbuilds.io/.
You can preview 1573c35 at https://pr46803-1573c35.ngbuilds.io/.

gkalpak commented on aio/content/guide/service-worker-config.md:
```diff
@@ -422,6 +432,12 @@ It is recommended that you use the default performance strategy whenever possibl
 
 </div>
 
+## Handling errors when the server can't be reached
```
This section feels a little out of place in this guide. I'd rather put it in the [SW introduction][1] or the [SW in production][2] guide.

[1]: https://pr46803-1573c35.ngbuilds.io/guide/service-worker-intro
[2]: https://pr46803-1573c35.ngbuilds.io/guide/service-worker-devops

bob-watson replied:
I agree.
How does it look in its new home?

mary-poppins commented:
You can preview e367391 at https://pr46803-e367391.ngbuilds.io/.

devversion committed:
feat(zone.js): include zone name when sync-test zone reports tasks

The sync-test zone is used in e.g. `describe` to raise an error when
there is asynchronous code scheduled in describe blocks. This commit
includes the zone name in such thrown errors to allow for us to include
the jasmine describe name in the error. This will be wired up in the
jasmine zonejs patches separately.

devversion committed:
feat(zone.js): include jasmine `describe` block name when raising unexpected task error

As mentioned in the previous commit that ensured that the Zone name is
included in errors raised by the `SyncTestZoneSpec`, we can now include
the Jasmine describe block descriptions in such errors.

Errors can often happen when users accidentally try to set up Angular
tests without an `it` block. Resulting in errors where it's not clear
at all which describe block (of potentially a large repository) is
involved:

```
 An error was thrown in afterAll
  error properties: Object({ originalStack: 'Error: Cannot call XX from within a sync test.
      at new ZoneAwareError (packages/zone.js/test/browser_test_rollup.umd.js:98:37)
      at e.onScheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:158:196)
      at e.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:7529)
      at t.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3539)
      at t.scheduleMicroTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3791)
      at r.execute (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:166:4372)
      at queueRunnerFa ...
      at <Jasmine>
```

We now include the describe block description in the error, so that it
is easier to figure out the location of the culprit code.

devversion committed:
fixup! feat(zone.js): include jasmine `describe` block name when raising unexpected task error

ds

devversion committed:
test: update tests to not run jasmine `done` function in sync-test zone from `describe` block

There are some ZoneJS tests that fork the zone from the `describe` block
for testing the zone patching. This does cause the Jasmine `done`
function later in `it` specs to be invoked in the sync-test zone from
the original `describe` block. The `done` implementation now has
changed with the Karma Jasmine update and breaks because it now causes
tasks to be scheduled.

It is conceptually incorrect/invalid to take the describe sync zone and
run test logic with that sync zone.

```
 An error was thrown in afterAll
  error properties: Object({ originalStack: 'Error: Cannot call jasmine.execute().forceTask from within a sync test (syncTestZone for jasmine.describe#FileReader).
      at new ZoneAwareError (packages/zone.js/test/browser_test_rollup.umd.js:98:37)
      at e.onScheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:158:196)
      at e.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:7529)
      at t.scheduleTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3539)
      at t.scheduleMicroTask (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:14:3791)
      at r.execute (packages/zone.js/bundles/zone-testing-bundle.umd.min.js:166:4312)
      at queueRunnerFa ...
      at <Jasmine>
```

thePunderWoman committed:
build: Add Jessica and Andrew K to docs-infra (#46843)

This adds two people to the docs-infra pullapprove group.

PR Close #46843

bob-watson commented on aio/content/guide/service-worker-config.md:
```diff
@@ -309,28 +322,28 @@ The Angular service worker can use either of two caching strategies for data res
-1.  If the network request does not complete after 0ms \(that is, immediately\), fall back to the cache \(ignoring cache age\).
-1.  Once the network request completes, update the cache for future requests.
-1.  If the resource does not exist in the cache, wait for the network request anyway.
+2.  If the network request does not complete immediately, ignore the cache age and fall back to the cached value.
```
Thanks!

bob-watson commented on aio/content/guide/service-worker-config.md:
```diff
@@ -309,28 +322,28 @@ The Angular service worker can use either of two caching strategies for data res
-| Groups with the `freshness` strategy   | The default value is `true` \(cache opaque responses\). These groups will request the data anew every time, only falling back to the cached response when offline or on a slow network. Therefore, it doesn't matter if the service worker caches an error response.                                    |
-| Groups with the `performance` strategy | The default value is `false` \(do not cache opaque responses\). These groups would continue to return a cached response until `maxAge` expires, even if the error was due to a temporary network or server issue. Therefore, it would be problematic for the service worker to cache an error response. |
+| Groups with the `freshness` strategy   | The default value is `true` and the service worker caches opaque responses. These groups will request the data every time and only fall back to the cached response when offline or on a slow network. Therefore, it doesn't matter if the service worker caches an error response.                                    |
+| Groups with the `performance` strategy | The default value is `false` and the service work doen't cache opaque responses. These groups would continue to return a cached response until `maxAge` expires, even if the error was due to a temporary network or server issue. Therefore, it would be problematic for the service worker to cache an error response. |
```
Thanks.

Commit:
docs: Move server timeout paragraph.

* Moves the service worker service timeout paragraph to a more logical topic.
* Removes line error from the more logical topic.

fixes #46445

Commit:
docs: Move server timeout paragraph.

* Moves the service worker service timeout paragraph to a more logical topic.
* Removes line error from the more logical topic.

fixes #46445

mary-poppins commented:
You can preview 7a6fc47 at https://pr46803-7a6fc47.ngbuilds.io/.

cexbrayat committed:
refactor(forms): simplify group builder function (#46844)

Applies the same logic that we have in the `control` function.

PR Close #46844

jnizet committed:
docs(core): document that destroyAfterEach is true by default (#46845)

PR Close #46845

cexbrayat committed:
fix(compiler-cli): improve the missingControlFlowDirective message (#46846)

The extended diagnostics about missing control flow directive was only mentioning that the `CommonModule` should be imported.
Now that the control flow directives are available as standalone, the message mentions that directive itself can be imported.

The message now also mentions which import should be used for the directive (as it can be tricky to figure out that `NgForOf` is the directive corresponding to `*ngFor`).

PR Close #46846

angular-robot committed:
build: update cross-repo angular dependencies to 84dd092 (#46848)

See associated pull request for more information.

PR Close #46848

mary-poppins commented:
You can preview d45ec84 at https://pr46803-d45ec84.ngbuilds.io/.

thePunderWoman committed:
Revert "build: update cross-repo angular dependencies to 84dd092 (#46848)" (#46849)

This reverts commit 510f45957a4fa43fbdec10f62451ae4c4aa6fa23.

PR Close #46849

thePunderWoman committed:
Revert "Revert "build: update cross-repo angular dependencies to 84dd092 (#46848)" (#46849)" (#46852)

This reverts commit b2a74469de7b29f8700b1673fd5302c30fc99afc.

PR Close #46852

Piyush132000 committed:
docs: fix code style (#46862)

PR Close #46862

dario-piotrowicz committed:
fix(animations): make sure falsy values are added to _globalTimelineStyles (#46863)

style values get added to the `_globalTimelineStyles` map in order to keep
them so that they can be used across different timelines

`_globalTimelineStyles` was previously a plain object but has been
refactored to a map in #44482, as part of the update a check has been
changed from a ternary operation to an or (||), causing falsy values (as 0)
not to be added to the map anymore, apply the nullish coalescing operator (??)
instead to make sure only `undefined` and `null` are filtered out

also since this aspect was clearly not covered by tests, add a new test
to ensure that such regression doesn't happen in the future

resolves #46833

PR Close #46863

thePunderWoman committed:
build: fix forms size tracking numbers (#46871)

This updates the goldens for size tracking.

PR Close #46871

gkalpak commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -119,27 +118,35 @@ Other reasons the Angular service worker might change the version of a running a
+The service worker either returns a cached response or sends the request to the server, depending on the state and configuration of the cache. 
+The service worker only caches responses to non-mutating requests, such as `GET` and `HEAD`.
+
+If the service worker doesn't receive a response, it returns a [504 Gateway Timeout](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504) status. A `504` status could be returned because the server is offline or the client was disconnected.
```
This is definitely simpler than my [previous suggestion][1] (:+1:), but there are a couple of points that are missing/conveyed incorrectly. It would be great if we could make these more clear:

- This SW behavior applies to all failed requests. I.e. not just when no response is received, but also when an error response is received (for example, a response with a 4xx or 5xx status code). The client' being offline or the server's being down are just two examples (of many possible causes).

- I'd prefer to give the 504 response as an example of the type of response that the SW would generate, not as a de facto behavior. I.e. I would put more emphasis on the fact that the SW will generate its own error response and what it tries to indicate, rather than the specific status code.
  The reason is that I would like to discourage people from treating the 504 response as an API, but rather build their apps so they can handle regular failure responses. This is better for the apps (since you cannot rely on the SW always being in effect) and a easier for us to potentially change the SW behavior in the future (while keeping the same intent).

I don't feel too strongly about these, but if we could somehow include them (without hurting simplicity/clarity) it would be great 🚀 

[1]: https://github.com/angular/angular/pull/46803#discussion_r921454991

bob-watson replied:
I tried to work that in. Let me know what you think.

atscott committed:
refactor(router): Update TitleStrategy to useFactory (#46876)

The implementation of the `DefaultTitleStrategy` was modeled after the
existing strategy patterns in the Router. These patterns were developed
before the `providedIn` syntax for injectables. We can simplify the
model a lot by providing the default in the factory of the abstract
class.

Note that the other strategy patterns aren't touched in this PR due to
how long they've existed. Because they have been there for such a long
time, it's possible there will need to be some adjustments to code
if/when they are refactored to do the same.

PR Close #46876

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -35,15 +35,15 @@ In this scenario, it is not valid to serve the old `index.html`, which calls `st
 <div class="alert is-helpful">
 
-`ngsw.json` is the manifest file that is generated at build time based on `ngsw-config.json`.
+`ngsw.json` is the manifest file that is created at build time based on `ngsw-config.json`.
```
simpler language.

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -52,25 +52,25 @@ With the versioning behavior of the Angular service worker, an application serve
-A service worker ignores such constraints and effectively long caches the entire application.
-Consequently, it is essential that the service worker gets the correct content.
+One of the potential side effects of long caching is inadvertently caching a resource that's not valid.
+In a normal HTTP cache, a hard refresh or the cache expiring limits the negative effects of caching an file that's not valid.
```
Thanks!

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -52,25 +52,25 @@ With the versioning behavior of the Angular service worker, an application serve
+One of the potential side effects of long caching is inadvertently caching a resource that's not valid.
+In a normal HTTP cache, a hard refresh or the cache expiring limits the negative effects of caching an file that's not valid.
+A service worker ignores such constraints and effectively long-caches the entire application.
+As a result, the service worker must get the correct content.
```
This section was hard for me to parse. I see a lot of facts, but what's important and what's merely descriptive is hard to separate. 

Re-reading this, it looks like we might be burying the lede, in that the critical fact here seems to be that **the service worker long-caches the application.**

Reading further, I'm thinking the **Resource integrity** section might benefit from a rewrite to help sort out descriptions from key takeaways.  Adding a section to indicate **Hashed content** as separate from **Unhashed content**, which already has a header, might also help.

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -52,25 +52,25 @@ With the versioning behavior of the Angular service worker, an application serve
-If necessary, the service worker enters a safe mode where requests fall back on the network, opting not to use its cache if the risk of serving invalid, broken, or outdated content is high.
+If a particular file fails validation, the Angular service worker attempts to re-fetch the content using a "cache-busting" URL parameter to prevent browser or intermediate caching.
+If that content also fails validation, the service worker considers the entire version of the application to not be valid and stops serving the application.
+If necessary, the service worker enters a safe mode where requests fall back on the network. The service worker doesn't use its cache if there's a high risk of serving content that is not valid, broken, or outdated.
```
Interesting point. I see what you mean. I'll  reorder them which should remove that risk of confusion.

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -119,27 +118,35 @@ Other reasons the Angular service worker might change the version of a running a
+
+### Service worker requests when the server can't be reached
+
+All requests pass through the service worker to the service unless
```
What does the service worker pass them through to?

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -119,27 +118,35 @@ Other reasons the Angular service worker might change the version of a running a
+### Service worker requests when the server can't be reached
+
+All requests pass through the service worker to the service unless
+the [service worker is explicitly bypassed](guide/service-worker-devops#bypassing-the-service-worker).
```
Thanks. I missed the forest for the tree :)

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -282,35 +289,39 @@ Such tools can be powerful when used properly, but there are a few things to kee
-Also included in the `@angular/service-worker` NPM package is a small script `safety-worker.js`, which when loaded will unregister itself from the browser and remove the service worker caches.
+<!-- vale Angular.Google_Acronyms = NO -->
+
+Also included in the `@angular/service-worker` NPM package is a small script `safety-worker.js`, which when loaded un-registers itself from the browser and remove the service worker caches.
```
Thanks!

bob-watson commented on aio/content/guide/service-worker-devops.md:
```diff
@@ -321,11 +332,11 @@ You might have already encountered the error `The script resource is behind a re
+If you setup a redirect from the old location, such as `example.com`, to the new location, `www.example.com` in this example, the worker stops working.
 Also, the redirect won't even trigger for users who are loading the site entirely from Service Worker.
-The old worker \(registered at `example.com`\) tries to update and sends requests to the old location `example.com` which get redirected to the new location `www.example.com` and create the error `The script resource is behind a redirect, which is disallowed`.
+The old worker, which was registered at `example.com`, tries to update and sends requests to the old location `example.com`. This request is redirected to the new location `www.example.com` and creates the error: `The script resource is behind a redirect, which is disallowed`.
```
Thanks!

bob-watson commented on aio/content/guide/service-worker-config.md:
```diff
@@ -309,28 +322,28 @@ The Angular service worker can use either of two caching strategies for data res
-| Groups with the `freshness` strategy   | The default value is `true` \(cache opaque responses\). These groups will request the data anew every time, only falling back to the cached response when offline or on a slow network. Therefore, it doesn't matter if the service worker caches an error response.                                    |
-| Groups with the `performance` strategy | The default value is `false` \(do not cache opaque responses\). These groups would continue to return a cached response until `maxAge` expires, even if the error was due to a temporary network or server issue. Therefore, it would be problematic for the service worker to cache an error response. |
+| Groups with the `freshness` strategy   | The default value is `true` and the service worker caches opaque responses. These groups will request the data every time and only fall back to the cached response when offline or on a slow network. Therefore, it doesn't matter if the service worker caches an error response.                                    |
+| Groups with the `performance` strategy | The default value is `false` and the service work doesn't cache opaque responses. These groups would continue to return a cached response until `maxAge` expires, even if the error was due to a temporary network or server issue. Therefore, it would be problematic for the service worker to cache an error response. |
```
Thanks!

Commit:
docs: update in response to comments

Commit:
Merge branch 'issue-46445' of https://github.com/bob-watson/angular into issue-46445

bob-watson commented:
The commit list got all out-of-whack for this, so I opened PR #46882 with the commits that address the last round of comments.
